File: src/macos/LanMessengerTests/PeerIDTests.swift
import XCTest
@testable import LanMessenger

// Conversations are filed by identity key, not by LAN address. See PeerID.swift
// and PROTOCOL.md → History.
//
// The migration is the part worth guarding: it runs once over history written
// under addresses, and the failure it exists to prevent — one person's messages
// shown as another's — is exactly what a careless guess at an ambiguous address
// would produce. So the tests are mostly about what it refuses to do.
final class PeerIDTests: XCTestCase {

    // Real-shaped keys: base64 of 32 bytes.
    private let dell = Data(repeating: 0x11, count: 32).base64EncodedString()
    private let ari  = Data(repeating: 0x22, count: 32).base64EncodedString()

    private func entry(_ id: String?, at t: Double, _ text: String = "x") -> MessageEntry {
        MessageEntry(sender: "p", text: text, incoming: true, timestamp: t,
                     messageId: id, status: "", readReceiptSent: false)
    }

    // MARK: - What an id is

    func testAKeyIsAnIdAndAnAddressIsNot() {
        XCTAssertTrue(PeerID.isKey(dell))
        XCTAssertFalse(PeerID.isKey("192.168.1.27"))
        XCTAssertFalse(PeerID.isKey("ip:192.168.1.27"))
        // Base64, but not 32 bytes: not a key.
        XCTAssertFalse(PeerID.isKey("AQIDBA=="))
        XCTAssertFalse(PeerID.isKey(""))
    }

    func testLegacyIdsRoundTripTheirAddress() {
        let id = PeerID.legacy(address: "192.168.1.27")
        XCTAssertEqual(id, "ip:192.168.1.27")
        XCTAssertTrue(PeerID.isLegacy(id))
        XCTAssertEqual(PeerID.legacyAddress(id), "192.168.1.27")
        XCTAssertNil(PeerID.legacyAddress(dell))
    }

    // MARK: - Resolving one name

    func testAnAddressOwnedByExactlyOneContactBecomesTheirKey() {
        let contacts = [PeerID.Contact(publicKeyB64: dell, lastIP: "192.168.1.27"),
                        PeerID.Contact(publicKeyB64: ari, lastIP: "192.168.1.31")]
        XCTAssertEqual(PeerID.resolve(legacyName: "192.168.1.27", contacts: contacts), dell)
        XCTAssertEqual(PeerID.resolve(legacyName: "192.168.1.31", contacts: contacts), ari)
    }

    /// Two contacts recorded at one address is the DHCP collision itself.
    /// Picking either would hand one person's thread to the other.
    func testAnAddressTwoContactsShareIsNotGuessed() {
        let contacts = [PeerID.Contact(publicKeyB64: dell, lastIP: "192.168.1.27"),
                        PeerID.Contact(publicKeyB64: ari, lastIP: "192.168.1.27")]
        XCTAssertEqual(PeerID.resolve(legacyName: "192.168.1.27", contacts: contacts),
                       "ip:192.168.1.27")
    }

    func testAnAddressNoContactOwnsIsKeptUnderItsAddress() {
        let contacts = [PeerID.Contact(publicKeyB64: dell, lastIP: "192.168.1.27")]
        XCTAssertEqual(PeerID.resolve(legacyName: "192.168.1.9", contacts: contacts),
                       "ip:192.168.1.9")
    }

    func testKeysAndLegacyIdsPassThrough() {
        let contacts = [PeerID.Contact(publicKeyB64: dell, lastIP: "192.168.1.27")]
        XCTAssertEqual(PeerID.resolve(legacyName: ari, contacts: contacts), ari)
        XCTAssertEqual(PeerID.resolve(legacyName: "ip:192.168.1.27", contacts: contacts),
                       "ip:192.168.1.27")
    }

    func testARelayPlaceholderResolvesToTheOneContactWithThatKeyPrefix() {
        let contacts = [PeerID.Contact(publicKeyB64: dell, lastIP: "192.168.1.27"),
                        PeerID.Contact(publicKeyB64: ari, lastIP: "192.168.1.31")]
        let name = "relay-\(dell.prefix(8))"
        XCTAssertEqual(PeerID.resolve(legacyName: name, contacts: contacts), dell)
        // Nobody with that prefix: kept, and under exactly the name the live
        // re-filing in AppModel looks for once that peer turns up.
        XCTAssertEqual(PeerID.resolve(legacyName: name, contacts: []),
                       PeerID.relayPlaceholder(forKey: dell))
    }

    // MARK: - Re-filing a whole history

    /// A contact that moved address had a bucket under each. Both land on the
    /// key and merge: de-duplicated, in time order, capped.
    func testBucketsThatLandOnOneKeyMerge() {
        let contacts = [PeerID.Contact(publicKeyB64: dell, lastIP: "192.168.1.27")]
        let history: [String: [MessageEntry]] = [
            "192.168.1.27": [entry("b", at: 2), entry("d", at: 4)],
            dell: [entry("a", at: 1), entry("b", at: 2), entry("c", at: 3)],
        ]
        let (result, moved) = PeerID.rekey(history: history, contacts: contacts, cap: 200)

        XCTAssertEqual(moved, ["192.168.1.27": dell])
        XCTAssertEqual(Array(result.keys), [dell])
        XCTAssertEqual(result[dell]?.map(\.messageId), ["a", "b", "c", "d"])
    }

    func testAMergeKeepsTheNewestWithinTheCap() {
        let contacts = [PeerID.Contact(publicKeyB64: dell, lastIP: "192.168.1.27")]
        let history: [String: [MessageEntry]] = [
            "192.168.1.27": [entry("a", at: 1), entry("c", at: 3)],
            dell: [entry("b", at: 2), entry("d", at: 4)],
        ]
        let (result, _) = PeerID.rekey(history: history, contacts: contacts, cap: 3)
        XCTAssertEqual(result[dell]?.map(\.messageId), ["b", "c", "d"])
    }

    /// Messages sharing a timestamp keep the order they arrived in, and
    /// entries with no id (older attachments) are never de-duplicated away.
    func testTheMergeIsStableAndKeepsIdlessEntries() {
        let merged = PeerID.merge([[entry(nil, at: 5, "first"), entry("x", at: 5, "second")],
                                   [entry(nil, at: 5, "third"), entry("x", at: 5, "dup")]],
                                  cap: 200)
        XCTAssertEqual(merged.map(\.text), ["first", "second", "third"])
    }

    /// A bucket with one source is left exactly as it was — re-sorting it
    /// could swap two messages that share a timestamp.
    func testASingleBucketIsNotReordered() {
        let contacts = [PeerID.Contact(publicKeyB64: dell, lastIP: "192.168.1.27")]
        let original = [entry("late", at: 9), entry("early", at: 1)]
        let (result, _) = PeerID.rekey(history: ["192.168.1.27": original],
                                       contacts: contacts, cap: 200)
        XCTAssertEqual(result[dell]?.map(\.messageId), ["late", "early"])
    }

    /// Runs at every load, so a migrated history must pass through untouched
    /// and report nothing moved — otherwise every launch rewrites the file.
    func testRekeyingIsIdempotent() {
        let contacts = [PeerID.Contact(publicKeyB64: dell, lastIP: "192.168.1.27")]
        let history: [String: [MessageEntry]] = [
            "192.168.1.27": [entry("a", at: 1)],
            "192.168.1.9": [entry("b", at: 2)],
        ]
        let (once, _) = PeerID.rekey(history: history, contacts: contacts, cap: 200)
        let (twice, moved) = PeerID.rekey(history: once, contacts: contacts, cap: 200)
        XCTAssertTrue(moved.isEmpty)
        XCTAssertEqual(twice.keys.sorted(), once.keys.sorted())
        XCTAssertEqual(twice[dell]?.map(\.messageId), ["a"])
        XCTAssertEqual(twice["ip:192.168.1.9"]?.map(\.messageId), ["b"])
    }

    func testListsAreRekeyedInOrderWithoutDuplicates() {
        let contacts = [PeerID.Contact(publicKeyB64: dell, lastIP: "192.168.1.27")]
        XCTAssertEqual(
            PeerID.rekey(list: ["192.168.1.27", "192.168.1.9", dell], contacts: contacts),
            [dell, "ip:192.168.1.9"])
    }

    // MARK: - HistoryStore.merge

    func testHistoryStoreMergesAPlaceholderIntoTheKey() {
        let placeholder = PeerID.relayPlaceholder(forKey: dell)
        defer {
            HistoryStore.shared.delete(peer: placeholder)
            HistoryStore.shared.delete(peer: dell)
        }
        HistoryStore.shared.append(entry: entry("relayed", at: 1), forPeer: placeholder)
        HistoryStore.shared.append(entry: entry("lan", at: 2), forPeer: dell)

        XCTAssertTrue(HistoryStore.shared.merge(from: placeholder, into: dell))
        XCTAssertEqual(HistoryStore.shared.entries(forPeer: dell).map(\.messageId),
                       ["relayed", "lan"])
        XCTAssertTrue(HistoryStore.shared.entries(forPeer: placeholder).isEmpty)
        XCTAssertFalse(HistoryStore.shared.merge(from: placeholder, into: dell),
                       "nothing left to move")
    }
}

// Unencrypted packets — typing, receipts, delete_message — only CLAIM a sender
// key. Filed by that claim alone they would need no connection from anywhere in
// particular, so a claim is accepted only from an address the key is known at.
@MainActor
final class ClaimedSenderBindingTests: XCTestCase {

    private let key = Data(repeating: 0x33, count: 32).base64EncodedString()
    private var savedBinding: ((String, String) -> Bool)?
    private var savedTyping: ((String, String, Bool) -> Void)?
    private var typed: [(peer: String, active: Bool)] = []

    override func setUp() async throws {
        savedBinding = MessagingService.shared.isBoundAddress
        savedTyping = MessagingService.shared.onTypingUpdate
        typed = []
        MessagingService.shared.onTypingUpdate = { [weak self] peer, _, active in
            self?.typed.append((peer, active))
        }
        let key = self.key
        MessagingService.shared.isBoundAddress = { k, ip in k == key && ip == "192.168.1.27" }
    }

    override func tearDown() async throws {
        MessagingService.shared.isBoundAddress = savedBinding
        MessagingService.shared.onTypingUpdate = savedTyping
        HistoryStore.shared.delete(peer: key)
    }

    private func typing(from sender: String, at ip: String) {
        let pkt = TypingPacket(type: "typing", active: true, sender: "Dell",
                               senderPublicKeyB64: sender, port: 54232)
        MessagingService.shared.handlePacket(.typing(pkt, senderIP: ip))
    }

    func testAClaimFromTheKeysAddressIsFiledUnderTheKey() {
        typing(from: key, at: "192.168.1.27")
        XCTAssertEqual(typed.map(\.peer), [key])
    }

    /// Somebody else on the LAN naming the Dell's key: dropped, not shown as
    /// the Dell typing.
    func testAClaimFromAnyOtherAddressIsDropped() {
        typing(from: key, at: "192.168.1.31")
        XCTAssertTrue(typed.isEmpty)
    }

    func testASenderKeyThatIsNotAKeyIsDropped() {
        MessagingService.shared.isBoundAddress = nil   // would accept anything
        typing(from: "192.168.1.27", at: "192.168.1.27")
        XCTAssertTrue(typed.isEmpty)
    }

    func testAnUnboundDeleteLeavesTheMessageAlone() {
        HistoryStore.shared.append(
            entry: MessageEntry(sender: "Dell", text: "keep me", incoming: true, timestamp: 1,
                                messageId: "del-unbound", status: "", readReceiptSent: false),
            forPeer: key)
        let pkt = ReceiptPacket(type: "delete_message", messageId: "del-unbound",
                                sender: "Dell", senderPublicKeyB64: key, port: 54232)
        MessagingService.shared.handlePacket(.delete(pkt, senderIP: "192.168.1.31"))

        let stored = HistoryStore.shared.entries(forPeer: key).first
        XCTAssertEqual(stored?.text, "keep me")
        XCTAssertEqual(stored?.deleted, false)
    }
}

File: src/windows-native/LanMessenger.Tests/PeerIdTests.cs
using LanMessenger.Core.Persistence;
using LanMessenger.Core.Protocol;
using LanMessenger.Core.Services;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Text.Json;

namespace LanMessenger.Tests;

// Conversations are filed by identity key, not by LAN address. See PeerId.cs and
// PROTOCOL.md -> History.
//
// The migration is the part worth guarding: it runs once over history written
// under addresses, and the failure it exists to prevent — one person's messages
// shown as another's — is exactly what a careless guess at an ambiguous address
// would produce. So the tests are mostly about what it refuses to do.
// Mirror of PeerIDTests.swift; peer_id_vector.json is asserted by both.
[TestClass]
public class PeerIdTests
{
    // Real-shaped keys: base64 of 32 bytes.
    private static readonly string Dell = Convert.ToBase64String(Enumerable.Repeat((byte)0x11, 32).ToArray());
    private static readonly string Ari  = Convert.ToBase64String(Enumerable.Repeat((byte)0x22, 32).ToArray());

    private static MessageEntry Entry(string? id, double t, string text = "x") => new()
    {
        Sender = "p", Text = text, Incoming = true, Timestamp = t, MessageId = id,
    };

    // MARK: - What an id is

    [TestMethod]
    public void AKeyIsAnIdAndAnAddressIsNot()
    {
        Assert.IsTrue(PeerId.IsKey(Dell));
        Assert.IsFalse(PeerId.IsKey("192.168.1.27"));
        Assert.IsFalse(PeerId.IsKey("ip:192.168.1.27"));
        // Base64, but not 32 bytes: not a key.
        Assert.IsFalse(PeerId.IsKey("AQIDBA=="));
        Assert.IsFalse(PeerId.IsKey(""));
        Assert.IsFalse(PeerId.IsKey(null));
    }

    [TestMethod]
    public void LegacyIdsRoundTripTheirAddress()
    {
        var id = PeerId.Legacy("192.168.1.27");
        Assert.AreEqual("ip:192.168.1.27", id);
        Assert.IsTrue(PeerId.IsLegacy(id));
        Assert.AreEqual("192.168.1.27", PeerId.LegacyAddress(id));
        Assert.IsNull(PeerId.LegacyAddress(Dell));
    }

    // MARK: - Resolving one name

    /// Two contacts recorded at one address is the DHCP collision itself.
    /// Picking either would hand one person's thread to the other.
    [TestMethod]
    public void AnAddressTwoContactsShareIsNotGuessed()
    {
        var contacts = new List<PeerId.Contact> { new(Dell, "192.168.1.27"), new(Ari, "192.168.1.27") };
        Assert.AreEqual("ip:192.168.1.27", PeerId.Resolve("192.168.1.27", contacts));
    }

    [TestMethod]
    public void AnAddressOwnedByExactlyOneContactBecomesTheirKey()
    {
        var contacts = new List<PeerId.Contact> { new(Dell, "192.168.1.27"), new(Ari, "192.168.1.31") };
        Assert.AreEqual(Dell, PeerId.Resolve("192.168.1.27", contacts));
        Assert.AreEqual(Ari, PeerId.Resolve("192.168.1.31", contacts));
    }

    // MARK: - Re-filing a whole history

    /// A bucket with one source is left exactly as it was — re-sorting it could
    /// swap two messages that share a timestamp.
    [TestMethod]
    public void ASingleBucketIsNotReordered()
    {
        var contacts = new List<PeerId.Contact> { new(Dell, "192.168.1.27") };
        var history = new Dictionary<string, List<MessageEntry>>
        {
            ["192.168.1.27"] = [Entry("late", 9), Entry("early", 1)],
        };
        var (result, _) = PeerId.Rekey(history, contacts, 200);
        CollectionAssert.AreEqual(new[] { "late", "early" }, result[Dell].Select(e => e.MessageId).ToList());
    }

    /// Messages sharing a timestamp keep the order they arrived in, and entries
    /// with no id (older attachments) are never de-duplicated away.
    [TestMethod]
    public void TheMergeIsStableAndKeepsIdlessEntries()
    {
        var merged = PeerId.Merge(
            [[Entry(null, 5, "first"), Entry("x", 5, "second")],
             [Entry(null, 5, "third"), Entry("x", 5, "dup")]], 200);
        CollectionAssert.AreEqual(new[] { "first", "second", "third" }, merged.Select(e => e.Text).ToList());
    }

    /// Runs at every load, so a migrated history must pass through untouched and
    /// report nothing moved — otherwise every launch rewrites the file.
    [TestMethod]
    public void RekeyingIsIdempotent()
    {
        var contacts = new List<PeerId.Contact> { new(Dell, "192.168.1.27") };
        var history = new Dictionary<string, List<MessageEntry>>
        {
            ["192.168.1.27"] = [Entry("a", 1)],
            ["192.168.1.9"]  = [Entry("b", 2)],
        };
        var (once, _) = PeerId.Rekey(history, contacts, 200);
        var (twice, moved) = PeerId.Rekey(once, contacts, 200);
        Assert.AreEqual(0, moved.Count);
        CollectionAssert.AreEquivalent(once.Keys.ToList(), twice.Keys.ToList());
        Assert.AreEqual("b", twice["ip:192.168.1.9"].Single().MessageId);
    }

    [TestMethod]
    public void HistoryStoreMergesAPlaceholderIntoTheKey()
    {
        var placeholder = PeerId.RelayPlaceholder(Dell);
        try
        {
            HistoryStore.Shared.Append(Entry("relayed", 1), placeholder);
            HistoryStore.Shared.Append(Entry("lan", 2), Dell);

            Assert.IsTrue(HistoryStore.Shared.Merge(placeholder, Dell));
            CollectionAssert.AreEqual(new[] { "relayed", "lan" },
                HistoryStore.Shared.Entries(Dell).Select(e => e.MessageId).ToList());
            Assert.AreEqual(0, HistoryStore.Shared.Entries(placeholder).Count);
            Assert.IsFalse(HistoryStore.Shared.Merge(placeholder, Dell), "nothing left to move");
        }
        finally
        {
            HistoryStore.Shared.Delete(placeholder);
            HistoryStore.Shared.Delete(Dell);
        }
    }

    // MARK: - The shared vector

    private static JsonElement Vector()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "peer_id_vector.json");
        if (!File.Exists(path)) path = "peer_id_vector.json";
        using var document = JsonDocument.Parse(File.ReadAllBytes(path));
        return document.RootElement.Clone();
    }

    /// The same person reads both machines' histories, so the two platforms must
    /// file the same legacy bucket under the same id.
    [TestMethod]
    public void TheSharedVectorResolvesIdenticallyOnBothPlatforms()
    {
        var v = Vector();
        var contacts = v.GetProperty("contacts").EnumerateArray()
            .Select(c => new PeerId.Contact(c.GetProperty("public_key_b64").GetString()!,
                                            c.GetProperty("last_ip").GetString()!))
            .ToList();

        foreach (var c in v.GetProperty("is_key").EnumerateArray())
        {
            var id = c.GetProperty("id").GetString()!;
            Assert.AreEqual(c.GetProperty("is_key").GetBoolean(), PeerId.IsKey(id), id);
        }
        foreach (var c in v.GetProperty("resolve").EnumerateArray())
        {
            var name = c.GetProperty("name").GetString()!;
            Assert.AreEqual(c.GetProperty("id").GetString(), PeerId.Resolve(name, contacts), name);
        }
        var placeholder = v.GetProperty("relay_placeholder");
        Assert.AreEqual(placeholder.GetProperty("id").GetString(),
                        PeerId.RelayPlaceholder(placeholder.GetProperty("key").GetString()!));

        var rekey = v.GetProperty("rekey");
        var history = rekey.GetProperty("history").EnumerateObject().ToDictionary(
            b => b.Name,
            b => b.Value.EnumerateArray()
                  .Select(e => Entry(e.GetProperty("message_id").GetString(), e.GetProperty("timestamp").GetDouble()))
                  .ToList());
        var (result, moved) = PeerId.Rekey(history, contacts, rekey.GetProperty("cap").GetInt32());

        var expectedMoved = rekey.GetProperty("moved").EnumerateObject().ToDictionary(p => p.Name, p => p.Value.GetString()!);
        CollectionAssert.AreEquivalent(expectedMoved.ToList(), moved.ToList());
        var expected = rekey.GetProperty("expected").EnumerateObject().ToList();
        Assert.AreEqual(expected.Count, result.Count);
        foreach (var bucket in expected)
            CollectionAssert.AreEqual(
                bucket.Value.EnumerateArray().Select(e => e.GetString()).ToList(),
                result[bucket.Name].Select(e => e.MessageId).ToList(), bucket.Name);

        var list = v.GetProperty("rekey_list");
        CollectionAssert.AreEqual(
            list.GetProperty("expected").EnumerateArray().Select(e => e.GetString()).ToList(),
            PeerId.Rekey(list.GetProperty("list").EnumerateArray().Select(e => e.GetString()!), contacts));
    }
}

// Unencrypted packets — typing, receipts, delete_message — only CLAIM a sender
// key. Filed by that claim alone they would need no connection from anywhere in
// particular, so a claim is accepted only from an address the key is known at.
[TestClass]
public class ClaimedSenderBindingTests
{
    private static readonly string Key = Convert.ToBase64String(Enumerable.Repeat((byte)0x33, 32).ToArray());
    private Func<string, string, bool>? _saved;

    [TestInitialize]
    public void Bind()
    {
        _saved = MessagingService.Shared.IsBoundAddress;
        MessagingService.Shared.IsBoundAddress = (k, ip) => k == Key && ip == "192.168.1.27";
        HistoryStore.Shared.Append(new MessageEntry
        {
            Sender = "Dell", Text = "keep me", Incoming = true, Timestamp = 1, MessageId = "del-claim",
        }, Key);
    }

    [TestCleanup]
    public void Restore()
    {
        MessagingService.Shared.IsBoundAddress = _saved;
        HistoryStore.Shared.Delete(Key);
    }

    private static void DeleteFrom(string ip) =>
        MessagingService.Shared.HandlePacket(new ValidatedDelete(new ReceiptPacket
        {
            Type = "delete_message", MessageId = "del-claim", Sender = "Dell",
            SenderPublicKeyB64 = Key, Port = 54232,
        }, ip));

    /// Somebody else on the LAN naming the Dell's key: refused, not applied to
    /// the Dell's thread.
    [TestMethod]
    public void AnUnboundDeleteLeavesTheMessageAlone()
    {
        DeleteFrom("192.168.1.31");
        var stored = HistoryStore.Shared.Entries(Key).Single();
        Assert.AreEqual("keep me", stored.Text);
        Assert.IsFalse(stored.Deleted);
    }

    [TestMethod]
    public void ABoundDeleteIsFiledUnderTheKey()
    {
        DeleteFrom("192.168.1.27");
        Assert.IsTrue(HistoryStore.Shared.Entries(Key).Single().Deleted);
    }
}
